jbpm-flow/src/main/java/org/jbpm/ruleflow/instance/RuleFlowProcessInstance.java

```diff
@@ -16,24 +16,94 @@
+import org.jbpm.process.instance.context.exception.ExceptionScopeInstance;
 import org.jbpm.ruleflow.core.RuleFlowProcess;
 import org.jbpm.workflow.core.node.StartNode;
 import org.jbpm.workflow.instance.NodeInstance;
 import org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl;
+import org.jbpm.workflow.instance.impl.queue.RunProcessInstance;
+import org.jbpm.workflow.instance.node.EventNodeInstanceInterface;
+import org.kie.api.runtime.process.EventListener;
 
 public class RuleFlowProcessInstance extends WorkflowProcessInstanceImpl {
 
     private static final long serialVersionUID = 510l;
-    
+
+    private transient RunProcessInstance runProcessInstance = new RunProcessInstance();
+
+    public RuleFlowProcessInstance() {
+    }
+
     public RuleFlowProcess getRuleFlowProcess() {
         return (RuleFlowProcess) getProcess();
     }
 
-    public void internalStart(String trigger) {
-    	StartNode startNode = getRuleFlowProcess().getStart(trigger);
-    	if (startNode != null) {
-    		((NodeInstance) getNodeInstance(startNode)).trigger(null, null);
-    	}
+    // STACKLESS
+
+    public void internalStart( String trigger ) {
+        StartNode startNode = getRuleFlowProcess().getStart(trigger);
+        if( startNode != null ) {
+            NodeInstance startNodeInstance = ((NodeInstance) createNodeInstance(startNode));
+            if( isStackless() ) {
+                runProcessInstance.addNodeInstanceTrigger(startNodeInstance, null, null);
+                executeProcessInstance(runProcessInstance);
+            } else {
+                startNodeInstance.trigger(null, null);
+            }
+        }
+    }
+
+    // purely adding instances to an EXECUTING action
+
+    @Override
+    public void addNodeInstanceTrigger( NodeInstance nodeInstance, NodeInstance from, String type ) {
+        runProcessInstance.addNodeInstanceTrigger(nodeInstance, from, type);
+    }
+
+    @Override
+    public void removeNodeInstanceTrigger( NodeInstance nodeInstance ) {
+        runProcessInstance.removeNodeInstanceTrigger(nodeInstance);
+    }
+
+    // TRIGGER EXECUTION
+
+    @Override
+    public void triggerCompletedAndExecute(EventNodeInstanceInterface eventNodeInstance) {
+        eventNodeInstance.triggerCompleted();
+        executeProcessInstance(runProcessInstance);
+    }
+
+    @Override
+    public void signalEventAndExecute( EventListener eventListener, String type, Object event ) {
+        runProcessInstance.addAndTriggerSignalEventAction(eventListener, type, event);
+        executeProcessInstance(runProcessInstance);
+    }
+
+    // TODO: use this instead of triggerCompletedAndExecute
+    public void executeQueueIfNotAlreadyExecuting() {
+        executeProcessInstance(runProcessInstance);
+    }
+
+
+    // DBG
+    private final boolean debug = true;
+
+    private void executeProcessInstance(RunProcessInstance runProcessInstance) {
+        if( debug ) {
+            runProcessInstance.execute();
```

Comment:
out of curiosity, how is this useful as it does something else than when not in debug?